js-bson: Failed to load c++ bson extension, using pure JS version

seem here is some err when build this module when is npm install moogoose.

here is  the builderror.log

In file included from ../ext/bson.cc:36:
In file included from ../ext/bson.h:17:
../node_modules/nan/nan.h:261:25: error: redefinition of '_NanEnsureLocal'
NAN_INLINE v8::Local<T> _NanEnsureLocal(v8::Local<T> val) {
                        ^
../node_modules/nan/nan.h:256:25: note: previous definition is here
NAN_INLINE v8::Local<T> _NanEnsureLocal(v8::Handle<T> val) {
                        ^
../node_modules/nan/nan.h:661:13: error: no member named 'smalloc' in namespace 'node'
    , node::smalloc::FreeCallback callback
      ~~~~~~^
../node_modules/nan/nan.h:672:12: error: no matching function for call to 'New'
    return node::Buffer::New(v8::Isolate::GetCurrent(), data, size);
           ^~~~~~~~~~~~~~~~~
/Users/leeshine/.node-gyp/4.1.0/include/node/node_buffer.h:31:40: note: candidate function not viable: no known conversion from 'uint32_t' (aka 'unsigned int') to 'enum encoding' for 3rd argument
NODE_EXTERN v8::MaybeLocal<v8::Object> New(v8::Isolate* isolate,
                                       ^
/Users/leeshine/.node-gyp/4.1.0/include/node/node_buffer.h:43:40: note: candidate function not viable: 2nd argument ('const char *') would lose const qualifier
NODE_EXTERN v8::MaybeLocal<v8::Object> New(v8::Isolate* isolate,
                                       ^
/Users/leeshine/.node-gyp/4.1.0/include/node/node_buffer.h:28:40: note: candidate function not viable: requires 2 arguments, but 3 were provided
NODE_EXTERN v8::MaybeLocal<v8::Object> New(v8::Isolate* isolate, size_t length);
                                       ^
/Users/leeshine/.node-gyp/4.1.0/include/node/node_buffer.h:36:40: note: candidate function not viable: requires 5 arguments, but 3 were provided
NODE_EXTERN v8::MaybeLocal<v8::Object> New(v8::Isolate* isolate,
                                       ^
In file included from ../ext/bson.cc:36:
In file included from ../ext/bson.h:17:
../node_modules/nan/nan.h:676:12: error: no viable conversion from 'v8::MaybeLocal<v8::Object>' to 'v8::Local<v8::Object>'
    return node::Buffer::New(v8::Isolate::GetCurrent(), size);
           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/leeshine/.node-gyp/4.1.0/include/node/v8.h:210:7: note: candidate constructor (the implicit copy constructor) not viable: no known conversion from 'v8::MaybeLocal<v8::Object>' to 'const v8::Local<v8::Object> &' for 1st argument
class Local {
      ^
/Users/leeshine/.node-gyp/4.1.0/include/node/v8.h:210:7: note: candidate constructor (the implicit move constructor) not viable: no known conversion from 'v8::MaybeLocal<v8::Object>' to 'v8::Local<v8::Object> &&' for 1st argument
class Local {
      ^
/Users/leeshine/.node-gyp/4.1.0/include/node/v8.h:214:13: note: candidate template ignored: could not match 'Local' against 'MaybeLocal'
  V8_INLINE Local(Local<S> that)
            ^
/Users/leeshine/.node-gyp/4.1.0/include/node/v8.h:326:13: note: candidate template ignored: could not match 'S *' against 'v8::MaybeLocal<v8::Object>'
  V8_INLINE Local(S* that)
            ^
In file included from ../ext/bson.cc:36:
In file included from ../ext/bson.h:17:
../node_modules/nan/nan.h:683:26: error: no member named 'Use' in namespace 'node::Buffer'
    return node::Buffer::Use(v8::Isolate::GetCurrent(), data, size);
           ~~~~~~~~~~~~~~^
../ext/bson.cc:825:17: warning: 'DecodeBytes' is deprecated: Use DecodeBytes(isolate, ...) [-Wdeprecated-declarations]
                ssize_t len = DecodeBytes(args[0], BINARY);
                              ^
/Users/leeshine/.node-gyp/4.1.0/include/node/node.h:318:32: note: 'DecodeBytes' has been explicitly marked deprecated here
                inline ssize_t DecodeBytes(
                               ^
/Users/leeshine/.node-gyp/4.1.0/include/node/node.h:66:42: note: expanded from macro 'NODE_DEPRECATED'
    __attribute__((deprecated(message))) declarator
                                         ^
../ext/bson.cc:833:3: warning: 'DecodeWrite' is deprecated: Use DecodeWrite(isolate, ...) [-Wdeprecated-declarations]
                DecodeWrite(data, len, args[0], BINARY);
                ^
/Users/leeshine/.node-gyp/4.1.0/include/node/node.h:331:32: note: 'DecodeWrite' has been explicitly marked deprecated here
                inline ssize_t DecodeWrite(char* buf,
                               ^
/Users/leeshine/.node-gyp/4.1.0/include/node/node.h:66:42: note: expanded from macro 'NODE_DEPRECATED'
    __attribute__((deprecated(message))) declarator
                                         ^
../ext/bson.cc:928:28: warning: 'Encode' is deprecated: Use Encode(isolate, ...) [-Wdeprecated-declarations]
                Local<Value> bin_value = Encode(serialized_object, object_size, BINARY)->ToString();
                                         ^
/Users/leeshine/.node-gyp/4.1.0/include/node/node.h:299:45: note: 'Encode' has been explicitly marked deprecated here
                inline v8::Local<v8::Value> Encode(
                                            ^
/Users/leeshine/.node-gyp/4.1.0/include/node/node.h:66:42: note: expanded from macro 'NODE_DEPRECATED'
    __attribute__((deprecated(message))) declarator
                                         ^
In file included from ../ext/bson.cc:14:
/Users/leeshine/.node-gyp/4.1.0/include/node/v8.h:221:5: error: assigning to 'v8::Primitive *volatile' from incompatible type 'v8::Value *'
    TYPE_CHECK(T, S);
    ^~~~~~~~~~~~~~~~
/Users/leeshine/.node-gyp/4.1.0/include/node/v8.h:180:37: note: expanded from macro 'TYPE_CHECK'
    *(static_cast<T* volatile*>(0)) = static_cast<S*>(0);      \
                                    ^ ~~~~~~~~~~~~~~~~~~
../node_modules/nan/nan.h:414:12: note: in instantiation of function template specialization 'v8::Local<v8::Primitive>::Local<v8::Value>' requested here
    return NanEscapeScope(NanNew(v8::Undefined(v8::Isolate::GetCurrent())));
           ^
../node_modules/nan/nan.h:398:30: note: expanded from macro 'NanEscapeScope'
# define NanEscapeScope(val) scope.Escape(_NanEnsureLocal(val))
                             ^
In file included from ../ext/bson.cc:14:
/Users/leeshine/.node-gyp/4.1.0/include/node/v8.h:221:5: error: assigning to 'v8::Boolean *volatile' from incompatible type 'v8::Value *'
    TYPE_CHECK(T, S);
    ^~~~~~~~~~~~~~~~
/Users/leeshine/.node-gyp/4.1.0/include/node/v8.h:180:37: note: expanded from macro 'TYPE_CHECK'
    *(static_cast<T* volatile*>(0)) = static_cast<S*>(0);      \
                                    ^ ~~~~~~~~~~~~~~~~~~
../node_modules/nan/nan.h:424:12: note: in instantiation of function template specialization 'v8::Local<v8::Boolean>::Local<v8::Value>' requested here
    return NanEscapeScope(NanNew(v8::True(v8::Isolate::GetCurrent())));
           ^
../node_modules/nan/nan.h:398:30: note: expanded from macro 'NanEscapeScope'
# define NanEscapeScope(val) scope.Escape(_NanEnsureLocal(val))
                             ^
In file included from ../ext/bson.cc:14:
/Users/leeshine/.node-gyp/4.1.0/include/node/v8.h:221:5: error: assigning to 'v8::Function *volatile' from incompatible type 'v8::Value *'
    TYPE_CHECK(T, S);
    ^~~~~~~~~~~~~~~~
/Users/leeshine/.node-gyp/4.1.0/include/node/v8.h:180:37: note: expanded from macro 'TYPE_CHECK'
    *(static_cast<T* volatile*>(0)) = static_cast<S*>(0);      \
                                    ^ ~~~~~~~~~~~~~~~~~~
../node_modules/nan/nan.h:1514:12: note: in instantiation of function template specialization 'v8::Local<v8::Function>::Local<v8::Value>' requested here
    return NanEscapeScope(NanNew(handle)->Get(kCallbackIndex)
           ^
../node_modules/nan/nan.h:398:30: note: expanded from macro 'NanEscapeScope'
# define NanEscapeScope(val) scope.Escape(_NanEnsureLocal(val))
                             ^
In file included from ../ext/bson.cc:14:
/Users/leeshine/.node-gyp/4.1.0/include/node/v8.h:221:5: error: assigning to 'v8::Object *volatile' from incompatible type 'v8::Value *'
    TYPE_CHECK(T, S);
    ^~~~~~~~~~~~~~~~
/Users/leeshine/.node-gyp/4.1.0/include/node/v8.h:180:37: note: expanded from macro 'TYPE_CHECK'
    *(static_cast<T* volatile*>(0)) = static_cast<S*>(0);      \
                                    ^ ~~~~~~~~~~~~~~~~~~
../node_modules/nan/nan.h:1632:12: note: in instantiation of function template specialization 'v8::Local<v8::Object>::Local<v8::Value>' requested here
    return NanEscapeScope(handle->Get(NanNew(key)).As<v8::Object>());
           ^
../node_modules/nan/nan.h:398:30: note: expanded from macro 'NanEscapeScope'
# define NanEscapeScope(val) scope.Escape(_NanEnsureLocal(val))
                             ^
3 warnings and 9 errors generated.
make: *** [Release/obj.target/bson/ext/bson.o] Error 1
gyp ERR! build error 
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/usr/local/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:270:23)
gyp ERR! stack     at emitTwo (events.js:87:13)
gyp ERR! stack     at ChildProcess.emit (events.js:172:7)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:200:12)
gyp ERR! System Darwin 14.5.0
gyp ERR! command "/usr/local/bin/node" "/usr/local/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /code/readbox/node_modules/mongoose/node_modules/mongodb/node_modules/bson
gyp ERR! node -v v4.1.0
gyp ERR! node-gyp -v v3.0.3
gyp ERR! not ok

--- a/.travis.yml
+++ b/.travis.yml
@@ -4,6 +4,7 @@ node_js:
   - "0.12"
   - "4"
   - "5"
+  - "6"
 sudo: false
 env:
   - CXX=g++-4.8

--- a/appveyor.yml
+++ b/appveyor.yml
@@ -1,0 +1,26 @@
+version: 0.0.{build}
+environment:
+  matrix:
+    - nodejs_version: "0.10"
+    - nodejs_version: "0.12"
+    - nodejs_version: "4"
+    - nodejs_version: "5"
+os:
+  - Visual Studio 2013
+  - Visual Studio 2015
+install:
+  - ps: Install-Product node $env:nodejs_version x64
+  # Sets Windows 7.1 SDK env vars.
+  - '"C:\Program Files\Microsoft SDKs\Windows\v7.1\Bin\SetEnv.cmd" /x64'
+  # Prepend 2015 and 2013 tools. Harmless when the path for the other version doesn't exist.
+  - set "PATH=%ProgramFiles(x86)%\MSBuild\14.0\Bin;%ProgramFiles(x86)%\MSBuild\12.0\Bin;%PATH%"
+  # Upgrade npm to latest
+  - npm install --loglevel error -g npm
+  - set "PATH=%APPDATA%\npm;%PATH%"
+  - node -v
+  - npm -v
+  - npm install -g --loglevel error node-gyp
+  - npm install
+build: off
+test_script:
+  - cmd: npm test

--- a/appveyor.yml
+++ b/appveyor.yml
@@ -5,6 +5,7 @@ environment:
     - nodejs_version: "0.12"
     - nodejs_version: "4"
     - nodejs_version: "5"
+    - nodejs_version: "6"
 os:
   - Visual Studio 2013
   - Visual Studio 2015

--- a/ext/bson.cc
+++ b/ext/bson.cc
@@ -366,7 +366,7 @@ template<typename T> void BSONSerializer<T>::SerializeValue(void* typeLocation, 
        Local<Object> buffer = ObjectWrap::Unwrap<Buffer>(value->ToObject());
 			 uint32_t length = object->length();
 	    #else
-			 uint32_t length = node::Buffer::Length(value->ToObject());
+			 uint32_t length = (uint32_t)node::Buffer::Length(value->ToObject());
 	    #endif
 
 			this->WriteInt32(length);
@@ -595,7 +595,8 @@ Local<Value> BSONDeserializer::DeserializeValue(BsonType type, bool promoteLongs
 					p -= 8;
 					// Read the 64 bit value
 					int64_t finalValue = (int64_t) ReadInt64();
-					return Nan::New<Number>(finalValue);
+					// JS has no int64 type.
+					return Nan::New<Number>(static_cast<double>(finalValue));
 				}
 			}
 
@@ -681,7 +682,7 @@ NAN_METHOD(BSON::New) {
 	Nan::HandleScope scope;
 
 	// Var maximum bson size
-	size_t maxBSONSize = MAX_BSON_SIZE;
+	uint32_t maxBSONSize = MAX_BSON_SIZE;
 
 	// Check that we have an array
 	if(info.Length() >= 1 && info[0]->IsArray()) {
@@ -919,7 +920,7 @@ NAN_METHOD(BSON::BSONSerialize) {
 	// If we have 3 arguments
 	if(info.Length() == 3 || info.Length() == 4) {
 		// NewBuffer takes ownership on the memory, so no need to free the final allocation
-		Local<Object> buffer = Unmaybe(Nan::NewBuffer(serialized_object, object_size));
+		Local<Object> buffer = Unmaybe(Nan::NewBuffer(serialized_object, (uint32_t)object_size));
 		info.GetReturnValue().Set(buffer);
 	} else {
 		Local<Value> bin_value = Nan::Encode(serialized_object, object_size)->ToString();

--- a/ext/bson.h
+++ b/ext/bson.h
@@ -19,6 +19,10 @@ const int _is_bigendian = 1;
 #include <v8.h>
 #include "nan.h"
 
+#if defined(_MSC_VER)
+#include <intrin.h>
+#endif
+
 using v8::Local;
 using v8::Value;
 using v8::Integer;
@@ -202,13 +206,21 @@ public:
 #if USE_MISALIGNED_MEMORY_ACCESS
 	void	WriteInt32(int32_t value)	{
 		if((size_t)((p + 4) - destinationBuffer) > MAX_BSON_SIZE) throw "document is larger than max bson document size of 16MB";
+#if defined (_MSC_VER)
+		*reinterpret_cast<int32_t*>(p) = is_bigendian() ? _byteswap_ulong(value) : value;
+#elif defined (__GNUC__)
 		*reinterpret_cast<int32_t*>(p) = is_bigendian() ? __builtin_bswap32(value) : value;
+#endif
 		p += 4;
 	}
 
 	void	WriteInt64(int64_t value)	{
 		if((size_t)((p + 8) - destinationBuffer) > MAX_BSON_SIZE) throw "document is larger than max bson document size of 16MB";
+#if defined (_MSC_VER)
+		*reinterpret_cast<int64_t*>(p) = is_bigendian() ? _byteswap_uint64(value) : value;
+#elif defined (__GNUC__)
 		*reinterpret_cast<int64_t*>(p) = is_bigendian() ? __builtin_bswap64(value) : value;
+#endif
 		p += 8;
 	}
 
@@ -216,7 +228,11 @@ public:
 		if((size_t)((p + 8) - destinationBuffer) > MAX_BSON_SIZE) throw "document is larger than max bson document size of 16MB";
 		*reinterpret_cast<double*>(p) = value;
 		if (is_bigendian()){
+#if defined (_MSC_VER)
+			*reinterpret_cast<uint64_t*>(p) = _byteswap_uint64(*reinterpret_cast<int64_t*>(p));
+#elif defined (__GNUC__)
 			*reinterpret_cast<uint64_t*>(p) = __builtin_bswap64(*reinterpret_cast<int64_t*>(p));
+#endif
 			value = *reinterpret_cast<double*>(p);
 		}
 		p += 8;
@@ -306,7 +322,11 @@ public:
 #if USE_MISALIGNED_MEMORY_ACCESS
 	void	CommitSize(void* beginPoint) {
 		int32_t value =  (int32_t) (p - (char*) beginPoint);
+#if defined(_MSC_VER)
+		*reinterpret_cast<int32_t*>(beginPoint) = is_bigendian() ? _byteswap_ulong(value) : value;
+#elif defined(__GNUC__)
 		*reinterpret_cast<int32_t*>(beginPoint) = is_bigendian() ? __builtin_bswap32(value) : value;
+#endif
 	}
 #else
 	void	CommitSize(void* beginPoint) {
@@ -364,40 +384,56 @@ public:
 
 	unsigned char	ReadByte()			{ return *reinterpret_cast<unsigned char*>(p++); }
 #if USE_MISALIGNED_MEMORY_ACCESS
-	int32_t			ReadInt32()			{ 
-		int32_t returnValue = *reinterpret_cast<int32_t*>(p); p += 4; 
-		if (is_bigendian()) { 
-			return __builtin_bswap32(returnValue); 
-		} else { 
-			return returnValue; 
-		} 
+	int32_t			ReadInt32()			{
+		int32_t returnValue = *reinterpret_cast<int32_t*>(p); p += 4;
+		if (is_bigendian()) {
+#if defined(_MSC_VER)
+			return _byteswap_ulong(returnValue);
+#elif defined(__GNUC__)
+			return __builtin_bswap32(returnValue);
+#endif
+		} else {
+			return returnValue;
+		}
 	}
-	uint32_t		ReadUInt32()		{ 
-		uint32_t returnValue = *reinterpret_cast<uint32_t*>(p); p += 4;  
-		if (is_bigendian()) { 
-			return __builtin_bswap32(returnValue); 
-		} else { 
-			return returnValue; 
-		} 
+	uint32_t		ReadUInt32()		{
+		uint32_t returnValue = *reinterpret_cast<uint32_t*>(p); p += 4;
+		if (is_bigendian()) {
+#if defined(_MSC_VER)
+			return _byteswap_ulong(returnValue);
+#elif defined(__GNUC__)
+			return __builtin_bswap32(returnValue);
+#endif
+		} else {
+			return returnValue;
+		}
 	}
-	int64_t			ReadInt64()			{ 
-		int64_t returnValue = *reinterpret_cast<int64_t*>(p); p += 8; 
-		if (is_bigendian()) { 
-			return __builtin_bswap64(returnValue); 
-		} else { 
-			return returnValue; 
-		} 
+	int64_t			ReadInt64()			{
+		int64_t returnValue = *reinterpret_cast<int64_t*>(p); p += 8;
+		if (is_bigendian()) {
+#if defined(_MSC_VER)
+			return _byteswap_uint64(returnValue);
+#elif defined(__GNUC__)
+			return __builtin_bswap64(returnValue);
+#endif
+		} else {
+			return returnValue;
+		}
 	}
-	double			ReadDouble()		{ 
+	double			ReadDouble()		{
 		double returnValue;
 		if (is_bigendian()) {
+#if defined(_MSC_VER)
+			uint64_t tmp = _byteswap_uint64(*reinterpret_cast<uint64_t*>(p));
+#elif defined(__GNUC__)
 			uint64_t tmp = __builtin_bswap64(*reinterpret_cast<uint64_t*>(p));
+#endif
 			returnValue = *reinterpret_cast<double*>(&tmp);
 		}else{
 			returnValue = *reinterpret_cast<double*>(p);
 		}
-    	p += 8;
-    	return returnValue;
+		p += 8;
+		return returnValue;
 	}
 #else
 	int32_t			ReadInt32()			{ int32_t returnValue; memcpy(&returnValue, p, 4); p += 4; return returnValue; }

--- a/package.json
+++ b/package.json
@@ -6,7 +6,7 @@
     "bson",
     "parser"
   ],
-  "version": "0.1.13",
+  "version": "0.1.14",
   "author": "Christian Amor Kvalheim <christkv@gmail.com>",
   "contributors": [],
   "repository": {

--- a/package.json
+++ b/package.json
@@ -19,7 +19,7 @@
   },
   "dependencies": {
     "bindings": "^1.2.1",
-    "nan": "~2.0.9"
+    "nan": "~2.3.5"
   },
   "devDependencies": {
     "nodeunit": "~0.9.0"

